feat: Add reproducible Debian package builds and distribution

MoeMahhouk opened the pull request:
## Issue Addressed


This pull request introduces workflows and configuration updates to automate the building, testing, and releasing of reproducible Debian packages for the Lighthouse project. Below are the key changes grouped by theme:

### New GitHub Actions Workflows

* **Debian Package Release Workflow**: Added `.github/workflows/release-deb.yml` to automate the building, testing, and releasing of reproducible Debian packages for multiple architectures (`x86_64` and `aarch64`). Includes steps for dry-run testing, artifact uploads, and checksum generation.

## Proposed Changes

- Added the required configuration for cargo-deb in lighthouse's Cargo.toml.
- Added Makefile targets to build the debian packages for different architectures and test their reproducibility
- Added CI workflow **release-deb.yml** that builds and pushes the debian packages into assets similar to release-reproducible.yml
- Added a simple **lighthouse.service** to be included in the generated lighthouse debian package

## Additional Info

This is an incremental work on https://github.com/sigp/lighthouse/pull/7614 and requires it to be merged first

close #7618

MoeMahhouk committed:
chore: refactor reproducible builds

MoeMahhouk committed:
Add release-reproducible CI workflow

MoeMahhouk committed:
Add reproducible-build workflow to catch regressions

MoeMahhouk committed:
fixing linting issues

MoeMahhouk committed:
Add necessary setup for reproducible debian packaging

MoeMahhouk committed:
Add necessary setup for reproducible debian packaging

MoeMahhouk committed:
add CI workflow to build and upload deb packages as assets

MoeMahhouk committed:
add CI workflow to build and upload deb packages as assets

MoeMahhouk committed:
merge the deb packaging workflow into release.yml to avoid conflicts

MoeMahhouk committed:
merge the deb packaging workflow into release.yml to avoid conflicts

bakhtin commented on lighthouse/lighthouse.service:
```diff
@@ -0,0 +1,34 @@
+[Service]
+Type=simple
+DynamicUser=true
+User=lighthouse
```
Both User and Group will most probably not exist on the target system. You should create it with [postinst maintainer scripts](https://github.com/kornelski/cargo-deb)

bakhtin commented on lighthouse/lighthouse.service:
```diff
@@ -0,0 +1,34 @@
+Type=simple
+DynamicUser=true
+User=lighthouse
+Group=eth
```
```suggestion
Group=lighthouse
```

MoeMahhouk replied:
I thought we wanted to pack, lighthouse, reth and rbuilder in the same group, right?

bakhtin replied:
We should not assume that either reth or rbuilder will be installed on the system. LH should be self-sufficient. Also, since you're using `DynamicUser` you can omit User and Group all together.

bakhtin replied:
I suggest to simplify the `[Service]` section to this:
```
[Service]
Type=exec # switched from 'simple' for better startup handling. See https://www.freedesktop.org/software/systemd/man/latest/systemd.service.html#Type=
DynamicUser=yes
StateDirectory=lighthouse
ExecStart=/usr/bin/lighthouse bn \
          --execution-endpoint http://localhost:8551 \
          --execution-jwt-secret-key 0000000000000000000000000000000000000000000000000000000000000000
          --datadir %S/lighthouse
```
This is unassuming, minimal, and still functional. An administrator can override the unit parameters if they want to.

MoeMahhouk replied:
good point. Adjusted as suggested

bakhtin commented on lighthouse/lighthouse.service:
```diff
@@ -0,0 +1,34 @@
+RestartSec=5
+TimeoutStopSec=180
+Environment="RUST_LOG=info"
+ExecStart=/usr/bin/lighthouse \
```
You should not assume any specific use-case of a service. I'd simplify it to `/usr/bin/lighthouse bn`

MoeMahhouk replied:
yeah, this is just a dummy/simple systemd service file that will be anyway overwritten during deplyoments as we discussed offline, right?
Otherwise, how would you suggest it to be? just` /usr/bin/lighthouse --help`?

bakhtin replied:
see above

bakhtin commented on lighthouse/lighthouse.service:
```diff
@@ -0,0 +1,34 @@
+
+[Service]
+Type=simple
+DynamicUser=true
```
```suggestion

```

bakhtin replied:
I'd suggest to rely on static users/groups as DynamicUser has [complex interaction](https://0pointer.net/blog/dynamic-users-with-systemd.html) with the stateful objects on disk (e.g., LH database)

MoeMahhouk replied:
good point, I will remove it

bakhtin commented on .github/workflows/reproducible-build.yml:
```diff
@@ -0,0 +1,188 @@
+name: reproducible-build
```
I'd avoid imposing a responsibility for verifying reproducibility to third-parties (Lighthouse team in this case). Those who have reproducibility requirements should independently verify it relying on their infra.

bakhtin replied:
The variance in the environment is not sufficient to say that the builds are reproducible. Check out how Debian independently verify reproducibility by building on different OSes with a variable set of tools.

MoeMahhouk replied:
I also followed here the steps of reth and how they are doing it here https://github.com/paradigmxyz/reth/blob/main/.github/workflows/reproducible-build.yml
However, good point on if the lighthouse team would want this to be as part of their workflow or not. I would leave it for them to decide and then we can re-iterate on what the best way to do it

bakhtin commented on .github/workflows/release-reproducible.yml:
```diff
@@ -0,0 +1,142 @@
+name: release-reproducible
```
Can we make reproducible builds a default for Lighthouse? Not an additional target requiring extra effort to support

MoeMahhouk replied:
I created a separate workflow for the reproducible container builds as separation of concerns similar to what the reth team did here https://github.com/paradigmxyz/reth/blob/main/.github/workflows/release-reproducible.yml

MoeMahhouk committed:
refinements and following packaging best practices

MoeMahhouk committed:
refinements and following packaging best practices

MoeMahhouk committed:
refine the reproducible build and test arm github runner

MoeMahhouk committed:
make release-reproducible verify reproducibility

MoeMahhouk committed:
add label trigger for reproducible builds

chong-he committed:
Merge branch 'unstable' into unstable

MoeMahhouk committed:
Update lighthouse.service

simplified the lighthouse systemd service

MoeMahhouk committed:
Merge branch 'unstable' into unstable

chong-he committed:
Merge branch 'unstable' into unstable

MoeMahhouk committed:
update rust version for github actions

MoeMahhouk committed:
addressing PR feedback

MoeMahhouk committed:
remove the release summary from docker-reproducible.yml

MoeMahhouk committed:
removed exposed ports from the Dockerfile.reproducible

MoeMahhouk committed:
add stable/unstable pushes to the docker-reproducible workflow

MoeMahhouk committed:
remove reproducible-builds workflow

MoeMahhouk committed:
Merge branch 'unstable' into unstable

MoeMahhouk committed:
remove unnecessary is_tag checks

MoeMahhouk committed:
manual workflow trigger is just for testing purposes only

MoeMahhouk committed:
Add reproducibility build for jemalloc-sys

MoeMahhouk committed:
chore: remove unnecessary echo

michaelsproul committed:
Remove redundant PROFILE in Makefile

MoeMahhouk committed:
Merge branch 'sigp:unstable' into unstable

MoeMahhouk committed:
Merge branch 'deb-packaging' of https://github.com/MoeMahhouk/lighthouse into deb-packaging

MoeMahhouk committed:
remove merge conflict duplicates and fix lint issues

MoeMahhouk committed:
remove testing script

MoeMahhouk committed:
lint cleanup

MoeMahhouk committed:
chore: remove unused makefile target

bakhtin commented on lighthouse/debian/postinst:
```diff
@@ -0,0 +1,25 @@
+#!/bin/sh
```
You don't need this file. Your systemd unit will automatically create a directory and a user.

MoeMahhouk replied:
ah right, those are leftovers from the previous commits where I didn't have the DynamicUser.
Good catch. I will remove now

bakhtin commented on lighthouse/debian/postrm:
```diff
@@ -0,0 +1,17 @@
+#!/bin/sh
```
This file also not needed

MoeMahhouk committed:
remove unnecessary scripts

MoeMahhouk committed:
chore: extract release reproducible binaries in their own workflow

bakhtin commented on .github/workflows/release-reproducible.yml:
```diff
@@ -0,0 +1,126 @@
+            echo "  - Release Suite: run $RELEASE_SUCCESS"
+            echo "  - docker-reproducible: run $DOCKER_SUCCESS"
+            echo "should_proceed=true" >> $GITHUB_OUTPUT
+          else
```
This block does not actually wait for the completion of another workflow that this workflow depends on, it checks the status and exits if not ready.
As per [GH docs](https://docs.github.com/en/actions/reference/workflows-and-actions/events-that-trigger-workflows#workflow_run) when multiple WFs specified in `workflows: [Release Suite, docker-reproducible]` only one needs to complete, the other could still be running when `release-reproducible` is triggered. So we need to actually wait for the completion periodically checking the status of a WF.

MoeMahhouk replied:
yes exactly, that is why I introduced the `should_proceed` variable which will only be true if both of those workflows ran and completed successfully.
Therefore, if only one ran and the other didnt, it will immediately stop this workflow to spare resources and run again once the second workflow finalizes successfully. Check out the types: [completed] [above](https://github.com/sigp/lighthouse/pull/7617/files#diff-222af23bee616920b04f5b92a83eb5106fce08abd885cd3a3b15b8beb5e789c3R10)

chong-he commented:
@MoeMahhouk is this still a work in progress?

MoeMahhouk commented:
> @MoeMahhouk is this still a work in progress?

@chong-he , yes it is. I will open it up for review this week after clearing up some feedback.
Then I would like to get your and your team's feedback on it to bring it to final shape

MoeMahhouk commented on Makefile:
```diff
@@ -332,3 +332,81 @@ clean:
+remove-deb-local: ## Remove installed lighthouse package
+	@echo "Removing lighthouse package..."
+	sudo dpkg -r lighthouse || true
+	sudo systemctl daemon-reload || true
```
I can remove those targets if not needed to minimize PR code

MoeMahhouk commented on Makefile:
```diff
@@ -332,3 +332,81 @@ clean:
+	@echo ""
+	@echo "Quick start:"
+	@echo "  make deb-cargo         - Build .deb for current RUST_TARGET"
+	@echo "  make install-deb-local - Test the package"
```
same here, I can clean this up if not needed

MoeMahhouk committed:
Merge branch 'unstable' into deb-packaging

MoeMahhouk committed:
remove unnecessary targets

chong-he commented:
cargo sort CI is failing, can you fix this? 

> > @MoeMahhouk is this still a work in progress?
> 
> @chong-he , yes it is. I will open it up for review this week after clearing up some feedback. Then I would like to get your and your team's feedback on it to bring it to final shape

cargo sort CI is failing, can you fix this?

MoeMahhouk committed:
fix the cargo sort issues

MoeMahhouk commented:
> cargo sort CI is failing, can you fix this?
> 
> > > @MoeMahhouk is this still a work in progress?
> > 
> > 
> > @chong-he , yes it is. I will open it up for review this week after clearing up some feedback. Then I would like to get your and your team's feedback on it to bring it to final shape
> 
> cargo sort CI is failing, can you fix this?

Done

MoeMahhouk committed:
Merge branch 'unstable' into deb-packaging

chong-he commented:
I have a look at this PR, from my understanding, this will include a `.deb` package in the release alongside with other binary files? 

I am wondering what is the use case here?

MoeMahhouk commented:
> I have a look at this PR, from my understanding, this will include a `.deb` package in the release alongside with other binary files?
> 
> I am wondering what is the use case here?

The use case here is that those reproducibly built debian packages can be utilized to accelerate workflows that require building images that includes lighthouse. For example, BuilderNet VM image. In mkosi, you could use debian packages during the image build process which would reduce the build time substantially. Beside that use case, there are also other use-cases for such users that prefer downloading and installing lighthouse using debian packages. In addition, the binary within is built reproducibly for security purposes.
There are probably other use-cases too but those that come to my mind now. Maybe @bakhtin could also share more here

bakhtin commented:
@chong-he I agree with what Moe described regarding the use-cases for a *.deb package. It also makes easier for people running nodes to install a *.deb package that comes with a systemd service definition baking in basic CLI options of Lighthouse simplifying managing the service.

MoeMahhouk committed:
Merge branch 'sigp:stable' into deb-packaging

MoeMahhouk committed:
Merge branch 'unstable' into deb-packaging

MoeMahhouk committed:
refactor and refinements

dapplion committed:
revert unrelated changes from stable merge

Removes changes from #7892 (beacon head monitor), #8748 (bootnode fix),
and #8749 (v8.1.0 release) that were inadvertently pulled in by merging
stable into this branch. These are not part of the deb-packaging work.

dapplion committed:
merge upstream unstable and remove unrelated changes

dapplion commented:
Cleaned up unrelated changes that were pulled into the PR diff by merging `stable` into this branch. The `stable` branch contained commits from #7892 (beacon head monitor), #8748 (bootnode fix), and #8749 (v8.1.0 release) that reached `unstable` through a different merge path, causing them to show up as part of this PR's diff.

Merged latest upstream `unstable` and reset all non-deb-packaging files to match, so the PR diff now only contains the intended changes:

- `.github/workflows/docker-reproducible.yml` — extract `.deb` artifacts
- `.github/workflows/release-reproducible.yml` — sign and publish workflow
- `Dockerfile.reproducible` — build `.deb` in Docker
- `Makefile` — deb packaging targets
- `lighthouse/Cargo.toml` — cargo-deb metadata
- `lighthouse/debian/lighthouse.service` — systemd unit
- `lighthouse/debian/prerm` — pre-removal script

niccoloraspa commented:
Could we get this merged before next release? We would really love to use `deb` packages 🙏

dapplion committed:
resolve merge conflict with unstable

michaelsproul committed:
Merge branch 'unstable' into deb-packaging

antondlr commented:
@MoeMahhouk sorry for the silence on this, but it's been on my mind, and I'd like to propose a different direction: see #9147 
I'd very much appreciate your input on this matter!

MoeMahhouk commented:
> @MoeMahhouk sorry for the silence on this, but it's been on my mind, and I'd like to propose a different direction: see #9147 I'd very much appreciate your input on this matter!

@antondlr No problems, thanks for revisiting this PR back.
I have reviewed your proposal PR and left you comments directly on it. 
If I see correctly, the proposal PR does a solid refactoring but it missed the purpose of this PR which is debian packaging. 
Is that not wanted anymore?
